scalafix-core/src/main/scala/scalafix/patch/Patch.scala

```diff
@@ -0,0 +1,186 @@
+package scalafix
+package patch
+
+import metaconfig._
+import scala.collection.immutable.Seq
+import scala.meta._
+import scala.meta.contrib._
+import scala.meta.internal.ast.Helpers._
+import scala.meta.tokens.Token
+import scala.meta.tokens.Token
+import scalafix.config._
+import scalafix.syntax._
+import scalafix.patch.TokenPatch.Add
+import scalafix.patch.TokenPatch.Remove
+import scalafix.patch.TreePatch.RenamePatch
+import scalafix.patch.TreePatch.Replace
+
+import difflib.DiffUtils
+import org.scalameta.logger
+
+sealed abstract class Patch {
+  // NOTE: potential bottle-neck, this might be very slow for large
+  // patches. We might want to group related patches and enforce some ordering.
+  def +(other: Patch): Patch =
+    if (this eq other) this
+    else {
+      (this, other) match {
+        case (InCtx(ap, ac, am), b @ InCtx(bp, bc, bm)) =>
+          if (ac ne bc) ???
+          else InCtx(ap + bp, ac, am.orElse(bm))
+        case (_, InCtx(p, ctx, m)) => InCtx(p + this, ctx, m)
+        case (InCtx(p, ctx, m), _) => InCtx(p + other, ctx, m)
+        case _ => Concat(this, other)
+      }
+    }
+  def ++(other: Seq[Patch]): Patch = other.foldLeft(this)(_ + _)
+
+  def appliedDiff: String = {
+    ???
+//    val original = ctx.tree.syntax
+//    val obtained = applied[T]
+//    Patch.unifiedDiff(original, obtained)
+  }
+
+  def applied: String = Patch(this)
+//    Patch.apply[T](this)
+}
+
+private[scalafix] case class Concat(a: Patch, b: Patch) extends Patch
+private[scalafix] case object EmptyPatch extends Patch
+private[scalafix] case class InCtx(patch: Patch,
+                                   ctx: RewriteCtx,
+                                   mirror: Option[Mirror])
+    extends Patch
```

Comment:
I've added a long comment why this was needed. It's subject to change in the future, but that should not affect the public api.